### DISCUSSION TITLE
Wiring Diagram for the JK-BMS

### DISCUSSION THREAD

**Discussion by jonadis** [Louisvdw/dbus-serialbattery#106] [Q&A] [answered]
Can anyone help me out with the wiring for a JK-BMS?
I know it goes from the Victron device to a USB-to-RS232 adapter but where does it go from there?
Currently I'm monitoring my JK-BMS via MODBUS in ESPHOME which just goes directly from the ESP to the pins on the JK-BMS, so I assume I need the JK-BMS RS-485 adapter, but is there anything else?  Does any one have the pinouts of that guy documented?

**Comment by Louisvdw** [comment]
You need to connect usin a USB-RS485 device. Here are the ports depending on the JKBMS model
![image](https://user-images.githubusercontent.com/26077665/157882859-9e32eaa6-c555-45ac-8a10-03e6325a4e1f.png)

**Comment by jonadis** [comment]
Thanks, 
I have the JK-BD6A17S8P so its like the first picture above.  I know the pin on the far left is GND based upon multimeter testing and I'm also able to monitor the BMS using this ESPHome project just fine: [https://github.com/syssi/esphome-jk-bms](https://github.com/syssi/esphome-jk-bms)
The middle two pins I'm not sure which is which but I've tried things both ways around.  The Victron branded RS-485 cables are a little spendy and a little hard to come by so I rolled the dice on this instead: [https://www.amazon.com/gp/product/B073RCSYX1/](https://www.amazon.com/gp/product/B073RCSYX1/)
It does show up in the OS but just the 232 chip.  My understanding is that there's a 232 to 485 converter after the 232 chip.
```
root@rpi4-5079b8:~# lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 003: ID 0403:6001 Future Technology Devices International, Ltd FT232 Serial (UART) IC
Bus 001 Device 002: ID 2109:3431 VIA Labs, Inc. Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

I'm fairly certain that I have the serial-battery driver installed correctly, because I also tested with a Daly BMS and it worked fine.

Does anyone happen to have the "lsusb" output of a working USB-485 adapter that's working with a JK-BMS.
Does anyone happen to know which of the middle two pins is which on the JK-BMS's RS-485 port?  I've only seen them labeled as RX & TX which doesn't line up with RS-485.

Thanks

**Reply by Louisvdw** [reply, marked as answer]
That connector should be fine. You will see RT232 on the software even though there is a RS485 step in there. 

The JKBMS RS485 pinout should be GND, TX, RX, Vbat. Be careful of the Vbat as that can be 50+V or more. If you get the RS485 cable from JKBMS it fix that for you, but if you use your own converter raw you might need to get a 5V source on the BMS (just check on of the other pins)

There are 2 tipes of convertion cables. Isolated cables like the Victron one has a physical isolation to block any current flowing from the one side to the other side. So the BMS voltages can never reach the GX controller in this case, and this is a good safety measure. Isolated cables need power on both sides. USB already power the one side, but you will need to supply power from the BMS side as well so you will need to connect the V+ as well.

Non isolated cables have none if this safety and thus would be cheaper. The power from the USB can also supply everything, so you only need a common ground and V+ is not connected. 

Gnd should always be connected.
Rx\Tx are the signal wires. It is safe to swop them to check if they are the wrong way around without anything breaking.
V+ Only connect this if you are using an isolated cable, and then to a 5V power supply.

**Comment by jonadis** [comment]
I finally started making some progress, but it looks like maybe the BMS is returning some information that isn't expected by the driver...
I'm seeing this in the logs:
```
@40000000625845de38e6cf84 WARNING:SerialBattery:Testing Jkbms
@40000000625845df00be0e4c WARNING:SerialBattery:Connection established to Jkbms
@40000000625845df02d44034 WARNING:SerialBattery:Battery connected to dbus from /dev/ttyUSB0
@40000000625845e00799daac Traceback (most recent call last):
@40000000625845e00799ee34   File "/data/etc/dbus-serialbattery/dbushelper.py", line 159, in publish_battery
@40000000625845e00799fdd4     self.publish_dbus()
@40000000625845e0079a05a4   File "/data/etc/dbus-serialbattery/dbushelper.py", line 178, in publish_dbus
@40000000625845e0079a1544     midpoint, deviation = self.battery.get_midvoltage()
@40000000625845e0079a1d14   File "/data/etc/dbus-serialbattery/battery.py", line 197, in get_midvoltage
@40000000625845e0079a2cb4     return midpoint, abs(1 - half1voltage/half2voltage)*100
@40000000625845e0079c6704 ZeroDivisionError: float division by zero
```
Any thoughts on where to go from here?

**Reply by Louisvdw** [reply]
Strange, it looks like you don't have cell voltages. Do you have cells connected to the BMS

**Reply by jonadis** [reply]
I don't, I just have the BMS booted up with an adjustable power supply.
If that's all there is to it, I'll slap a battery together and try again.
Thanks

**Reply by Louisvdw** [reply]
It reads those voltages and sum the values and because the sum is zero we get the device by zero error. 
If you have cells connected it should read the correct values and then it will be fine, so just connect it up

**Reply by merlep** [reply]
Victron report that their RS485-USB is non isolated . See point #10.   https://www.victronenergy.com/live/energy-meters:start
CAn you provide an alternative  reference confirming that it is isloated ?

**Reply by herrep** [reply]
It is also my understanding that Victron's RS485-USB is non-isolated. However, what Iassume is that the JKBMS RS485 Interface is isolated according to this [German post](http://forum.drbacke.de/viewtopic.php?t=5192&start=73).

**Reply by merlep** [reply]
Thanks for that clarification ,
